src/garage/sampler/ray_batched_sampler.py

```diff
@@ -0,0 +1,216 @@
+function, and a rollout function
+"""
+from collections import defaultdict
+import pickle
+
+import numpy as np
+import psutil
+import ray
+
+from garage.experiment import deterministic
+from garage.misc import tensor_utils
+from garage.misc.prog_bar_counter import ProgBarCounter
+from garage.sampler.base import BaseSampler
+
+
+class RaySampler(BaseSampler):
+    """Collects Policy Rollouts in a data parallel fashion.
+
+    Args:
+        - algo: A garage algo object
+        - env: A gym/akro env object
+        - should_render(bool): should the sampler render the trajectories
+        - sampler_worker_cls: If none, uses the default SamplerWorker
+            class
+
+    """
+
+    def __init__(self,
+                 algo,
+                 env,
+                 seed,
+                 should_render=False,
+                 num_processors=None,
+                 sampler_worker_cls=None):
+        self._sampler_worker = ray.remote(SamplerWorker if sampler_worker_cls
+                                          is None else sampler_worker_cls)
+        self._env = env
+        self._algo = algo
+        self._seed = seed
+        deterministic.set_seed(self._seed)
+        self._max_path_length = self._algo.max_path_length
+        self._should_render = should_render
+        if not ray.is_initialized():
+            ray.init(ignore_reinit_error=True)
+        self._num_workers = num_processors if num_processors \
+            else psutil.cpu_count(logical=False)
+        self._all_workers = defaultdict(None)
+
+    def start_worker(self):
+        """Initialize a new ray worker."""
+        env_pkl_id = ray.put(pickle.dumps(self._env))
+        agent_pkl_id = ray.put(pickle.dumps(self._algo.policy))
+        for worker_id in range(self._num_workers):
+            self._all_workers[worker_id] = (self._sampler_worker.remote(
+                worker_id, env_pkl_id, agent_pkl_id, self._seed,
+                self._max_path_length, self._should_render))
+        self._idle_worker_ids = list(range(self._num_workers))
+
+    def obtain_samples(self, itr, num_samples):
+        """Sample the policy for new trajectories.
+
+        Args:
+            - itr(int): iteration number
+            - num_samples(int):number of steps the the sampler should collect
+        """
+        self._active_workers = []
+        self._active_worker_ids = []
+        pbar = ProgBarCounter(num_samples)
+        completed_samples = 0
+        traj = []
+        updating_workers = []
+        self._idle_worker_ids = list(range(self._num_workers))
+        curr_policy_params = self._algo.policy.get_param_values()
+        params_id = ray.put(curr_policy_params)
+        while self._idle_worker_ids:
+            worker_id = self._idle_worker_ids.pop()
+            worker = self._all_workers[worker_id]
+            updating_workers.append(worker.set_agent.remote(params_id))
+        while completed_samples < num_samples:
```

Comment:
perhaps can you break this logic into some private helper functions? it's a bit hard to follow if someone is just scanning.

splitting up the code into some logical blocks (i.e. with vertical whitespace) and adding comments would also help.

this is the heart of this class. it should be easy to read/understand.

Reply:
I can do that sure thing!

Reply:
resolved